Azure HA Teleport deployment guide (#29962)
gravitational/teleport: docs/pages/deploy-a-cluster/helm-deployments/azure.mdx

    --name aks --audience api://AzureADTokenExchange \
    --subject system:serviceaccount:<Var name="teleport-ns" />:<Var name="teleport-release" /> \
    --issuer "$(az aks show --resource-group <Var name="aks-rg" /> --name <Var name="aks-name" /> --query oidcIssuerProfile.issuerUrl -o tsv)""
```

## Step 3/5. Set values to configure the cluster

<Details title="License Secret" scopeOnly={false} scope={"enterprise"} opened={true}>

Before you can install Teleport Enterprise in your Kubernetes cluster, you will need to
create a secret that contains your Teleport license information.

(!docs/pages/includes/enterprise/obtainlicense.mdx!)

Create a secret from your license file. Teleport will automatically discover
this secret as long as your file is named `license.pem`.

```code
$ kubectl create namespace <Var name="teleport-ns" />
$ kubectl -n <Var name="teleport-ns" /> create secret generic license --from-file=license.pem
```

</Details>

Now we'll configure the `teleport-cluster` Helm chart to use the `azure` mode. Create a
file called `azure-values.yaml` containing the values you've selected above:

<ScopedBlock scope={["oss"]}>

> ```code

Reviewer: ```suggestion
```yaml
```

(Assuming this renders OK with the code blocks, it's way more readable and breaks the page up)

Author: `yaml` doesn't support `<Var>` tho :(

$ cat >azure-values.yaml <<EOF
chartMode: azure
# Name of your cluster. Use the FQDN you intend to configure in DNS later
clusterName: teleport.example.com
azure:
  databaseHost: <Var name="teleport-pg" />.postgres.database.azure.com
  databaseUser: <Var name="pguser" />
  sessionRecordingStorageAccount: <Var name="teleportblob" />.blob.core.windows.net
  # Whether to mirror audit log entries to stdout in JSON format (useful for external log collectors)
  auditLogMirrorOnStdout: false

>   clientID: $(az identity show --resource-group <Var name="teleport-rg" /> --name teleport-id --query clientId -o tsv)

Reviewer: It's probably providing people with a way to get and then insert this themselves, as many people who read these guides don't understand what a heredoc is or how it works.

Author: Done in 11af5ab

highAvailability:
  # Number of replicas to configure
  replicaCount: 2
  certManager:
    # Enable cert-manager support to get TLS certificates
    enabled: true
    # Name of the cert-manager GlobalIssuer or Issuer to use
    issuerName: letsencrypt-production

>     issuerKind: GlobalIssuer

Reviewer: Might be worth mentioning that you'll need to configure `cert-manager` indepedently.

Author: The "prerequisites" section has this already 😅
> You will also need an Azure DNS zone, and access to an AKS cluster with cert-manager
[installed](https://cert-manager.io/docs/installation/) and configured to [issue certificates for said Azure DNS zone](https://cert-manager.io/docs/configuration/acme/dns01/azuredns/).

highAvailability:
  # Number of replicas to configure
  replicaCount: 2
  certManager:
    # Enable cert-manager support to get TLS certificates
    enabled: true
    # Name of the cert-manager GlobalIssuer or Issuer to use
    issuerName: letsencrypt-production
    issuerKind: GlobalIssuer
# If you are running Kubernetes 1.23 or above, disable PodSecurityPolicies
podSecurityPolicy:
  enabled: false
# Indicate that this is a Teleport Enterprise deployment
enterprise: true
EOF
```

</ScopedBlock>

Install the chart with the values from your `azure-values.yaml` file using this command:

```code
$ helm install <Var name="teleport-release" /> teleport/teleport-cluster \
    --create-namespace --namespace <Var name="teleport-ns" /> \
    --values azure-values.yaml
```

<Admonition type="note">
  You cannot change the `clusterName` after the cluster is configured, so make sure you choose wisely. We recommend using the fully-qualified domain name that you'll use for external access to your Teleport cluster.
</Admonition>

Once the chart is installed, you can use `kubectl` commands to view the deployment:

```code

> $ kubectl --namespace teleport get all

Reviewer: ```suggestion
$ kubectl --namespace <Var name="teleport-ns" /> get all
```

service/teleport            LoadBalancer   10.40.11.180   34.138.177.11   443:30258/TCP,3023:31802/TCP,3026:32182/TCP,3024:30101/TCP,3036:30302/TCP   22h
service/teleport-auth       ClusterIP      10.40.8.251    <none>          3025/TCP,3026/TCP                                                           22h
service/teleport-auth-v13   ClusterIP      None           <none>          <none>                                                                      22h
service/teleport-auth-v14   ClusterIP      None           <none>          <none>                                                                      22h

NAME                             READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/teleport-auth    2/2     2            2           22h
deployment.apps/teleport-proxy   2/2     2            2           22h

NAME                                       DESIRED   CURRENT   READY   AGE
replicaset.apps/teleport-auth-57989d4cb    2         2         2       22h
replicaset.apps/teleport-proxy-c6bf55cfc   2         2         2       22h
```

## Step 4/5. Set up DNS

> We'll now set up DNS `A` records for `teleport.example.com` and `*.teleport.example.com`.

Reviewer: I'd add something about how this assumes that you're using a domain configured with Azure DNS, and to add the records in your own provider if you're not.

Author: 519d9fcda1

```code
# Change these parameters if you altered them above

> $ NAMESPACE=teleport
> $ RELEASE_NAME=teleport

Reviewer: ```suggestion
$ NAMESPACE=<Var name="teleport-ns" /> 
$ RELEASE_NAME=<Var name="teleport-release" />
```

$ PREFIX=teleport
$ ZONE=example.com
$ ZONE_RG=dns-rg
$ external_ip="$(kubectl --namespace ${NAMESPACE} get service/${RELEASE_NAME} -o jsonpath='{.status.loadBalancer.ingress[*].ip}')"
$ az network dns record-set a add-record --resource-group ${ZONE_RG} --zone-name ${ZONE} --record-set-name ${PREFIX} --ipv4-address "${external_ip}"
$ az network dns record-set a add-record --resource-group ${ZONE_RG} --zone-name ${ZONE} --record-set-name "*.${PREFIX}"" --ipv4-address "${external_ip}"
```

## Step 5/5. Create a Teleport user

Create a user to be able to log into Teleport. This needs to be done on the Teleport auth server,
so we can run the command using `kubectl`:

<ScopedBlock scope={["oss"]}>
```code

> $ kubectl --namespace teleport exec deploy/teleport-auth -- tctl users add test --roles=access,editor

Reviewer: ```suggestion
$ kubectl --namespace <Var name="teleport-ns" /> exec deploy/<Var name="teleport-release" />-auth -- tctl users add test --roles=access,editor
```

User "test" has been created but requires a password. Share this URL with the user to complete user setup, link is valid for 1h:
https://teleport.example.com:443/web/invite/91cfbd08bc89122275006e48b516cc68

NOTE: Make sure teleport.example.com:443 points at a Teleport proxy that users can access.
```
</ScopedBlock>

> <ScopedBlock scope={["enterprise", "cloud"]}>

Reviewer: This is a cluster deployment guide, so doesn't need the Cloud scope

Author: In my defense, I copypasted this from one of the other guides LOL

Author: Switched to `TabItem`s scoped for oss and enterprise.

```code

> $ kubectl --namespace teleport exec deploy/teleport-auth -- tctl users add test --roles=access,editor,reviewer

Reviewer: ```suggestion
$ kubectl --namespace <Var name="teleport-ns" /> exec deploy/<Var name="teleport-release" />-auth -- tctl users add test --roles=access,editor,reviewer
```

Load the user creation link to create a password and set up 2-factor authentication for the Teleport user via the web UI.

### High Availability

In this guide, we have configured 2 replicas. This can be changed after cluster creation by altering the `highAvailability.replicaCount`
value [using `helm upgrade` as detailed below](#upgrading-the-cluster-after-deployment).

## Upgrading the cluster after deployment

To make changes to your Teleport cluster after deployment, you can use `helm upgrade`.

Helm defaults to using the latest version of the chart available in the repo, which will also correspond to the latest
version of Teleport. You can make sure that the repo is up to date by running `helm repo update`.

If you want to use a different version of Teleport, set the [`teleportVersionOverride`](../../reference/helm-reference/teleport-cluster.mdx#teleportversionoverride) value.

Here's an example where we set the chart to use 3 replicas:

<Tabs>
  <TabItem label="Using values.yaml">
  Edit your `azure-values.yaml` file from above and make the appropriate changes.

  Upgrade the deployment with the values from your `azure-values.yaml` file using this command:

  ```code

>   $ helm upgrade teleport teleport/teleport-cluster \
>     --namespace teleport \

Reviewer: ```suggestion
  $ helm upgrade <Var name="teleport-release" /> teleport/teleport-cluster \
    --namespace <Var name="teleport-ns" /> \
```

    -f azure-values.yaml
  ```

  </TabItem>
  <TabItem label="Using --set via CLI">
  Run this command, editing your command line parameters as appropriate:

  ```code

>   $ helm upgrade teleport teleport/teleport-cluster \
>     --namespace teleport \

Reviewer: ```suggestion
  $ helm upgrade <Var name="teleport-release" /> teleport/teleport-cluster \
    --namespace <Var name="teleport-ns" /> \
```

    --set highAvailability.replicaCount=3
  ```
  </TabItem>
</Tabs>

<Admonition type="note">
  To change `chartMode`, `clusterName` or any `azure` settings, you must first uninstall the existing chart and then install
  a new version with the appropriate values.
</Admonition>

## Uninstalling Teleport

To uninstall the `teleport-cluster` chart, use `helm uninstall <release-name>`. For example:

```code

> $ helm --namespace teleport uninstall teleport

Reviewer: ```suggestion
$ helm --namespace <Var name="teleport-ns" /> uninstall <Var name="teleport-release" />
```

First get the client ID for the `teleport-id` identity:
```code
$ az identity show --resource-group <Var name="teleport-rg" /> --name teleport-id --query clientId -o tsv
<Var name="teleport-clie-ntid-uuid-123456789012" />
```

Then create a file called `azure-values.yaml` containing the values you've selected above:

<Tabs>
<TabItem scope={["oss"]} label="Open Source">

```yaml
chartMode: azure
# Name of your cluster. Use the FQDN you intend to configure in DNS later
clusterName: teleport.example.com
azure:
  databaseHost: "<Var name="teleport-pg" />.postgres.database.azure.com"
  databaseUser: "<Var name="pguser" />"
  sessionRecordingStorageAccount: "<Var name="teleportblob" />.blob.core.windows.net"
  # Whether to mirror audit log entries to stdout in JSON format (useful for external log collectors)
  auditLogMirrorOnStdout: false
  clientID: "<Var name="teleport-clie-ntid-uuid-123456789012" />"
highAvailability:
  # Number of replicas to configure
  replicaCount: 2
  certManager:
    # Enable cert-manager support to get TLS certificates
    enabled: true
    # Name of the cert-manager GlobalIssuer or Issuer to use
    issuerName: letsencrypt-production

>     issuerKind: GlobalIssuer

Reviewer: I think this should be `ClusterIssuer` as per https://cert-manager.io/docs/concepts/issuer/?

```suggestion
    issuerKind: ClusterIssuer
```

Author: It should!

# If you are running Kubernetes 1.23 or above, disable PodSecurityPolicies
podSecurityPolicy:
  enabled: false
```

</TabItem>
<TabItem scope={["enterprise"]} label="Enterprise">

```yaml
chartMode: azure
# Name of your cluster. Use the FQDN you intend to configure in DNS later
clusterName: teleport.example.com
azure:
  databaseHost: "<Var name="teleport-pg" />.postgres.database.azure.com"
  databaseUser: "<Var name="pguser" />"
  sessionRecordingStorageAccount: "<Var name="teleportblob" />.blob.core.windows.net"
  # Whether to mirror audit log entries to stdout in JSON format (useful for external log collectors)
  auditLogMirrorOnStdout: false
  clientID: "<Var name="teleport-clie-ntid-uuid-123456789012" />"
highAvailability:
  # Number of replicas to configure
  replicaCount: 2
  certManager:
    # Enable cert-manager support to get TLS certificates
    enabled: true
    # Name of the cert-manager GlobalIssuer or Issuer to use
    issuerName: letsencrypt-production

>     issuerKind: GlobalIssuer

Reviewer: Ditto above

```suggestion
    issuerKind: ClusterIssuer
```

$ kubectl --namespace <Var name="teleport-ns" /> exec deploy/<Var name="teleport-release" />-auth -- tctl users add test --roles=access,editor,reviewer

User "test" has been created but requires a password. Share this URL with the user to complete user setup, link is valid for 1h:
https://teleport.example.com:443/web/invite/91cfbd08bc89122275006e48b516cc68

NOTE: Make sure teleport.example.com:443 points at a Teleport proxy that users can access.
```

</TabItem>
</Tabs>

Load the user creation link to create a password and set up 2-factor authentication for the Teleport user via the web UI.

### High Availability

In this guide, we have configured 2 replicas. This can be changed after cluster creation by altering the `highAvailability.replicaCount`
value [using `helm upgrade` as detailed below](#upgrading-the-cluster-after-deployment).

## Upgrading the cluster after deployment

To make changes to your Teleport cluster after deployment, you can use `helm upgrade`.

Helm defaults to using the latest version of the chart available in the repo, which will also correspond to the latest
version of Teleport. You can make sure that the repo is up to date by running `helm repo update`.

> If you want to use a different version of Teleport, set the [`teleportVersionOverride`](../../reference/helm-reference/teleport-cluster.mdx#teleportversionoverride) value.

Reviewer: We should advise people to use `--version` instead to keep the deployed chart version in sync with the deployed Teleport version.

```suggestion
If you want to use a different version of Teleport, pass the `--version` argument to Helm:

```code
$ helm upgrade --version 14.0.0 \
  <Var name="teleport-release" /> teleport/teleport-cluster \
  --namespace <Var name="teleport-ns" /> \
  -f azure-values.yaml
```

---
title: Running an HA Teleport cluster using Microsoft Azure, AKS, and Helm
description: Install and configure an HA Teleport cluster using a Microsoft Azure AKS cluster.
---

In this guide, we'll go through how to set up a High Availability Teleport
cluster with multiple replicas in Kubernetes using Teleport Helm charts and
Microsoft Azure managed services (Kubernetes Services, Database for PostgreSQL,
Blob Storage).

<Admonition type="tip" title="Have an existing Teleport cluster?">
If you are already running Teleport on another platform, you can use your
existing Teleport deployment to access your Kubernetes cluster. [Follow our
guide](../../kubernetes-access/getting-started.mdx) to connect your Kubernetes
cluster to Teleport.
</Admonition>

(!docs/pages/includes/cloud/call-to-action.mdx!)

## Prerequisites

(!docs/pages/kubernetes-access/helm/includes/teleport-cluster-prereqs.mdx!)

In addition, you will need `azure-cli` 2.51 or later to follow along these
instructions. Reference the Azure docs on [how to install the Azure
CLI](https://learn.microsoft.com/cli/azure/install-azure-cli).

> After installing it, make sure you are logged in, by typing `az login`. This

Reviewer 2: ```suggestion
After installing it, make sure you are logged in by typing `az login`. This
```
